thirdparty/patchmatchnet/models/net.py

```diff
@@ -0,0 +1,373 @@
+                )
+            else:
+                depth, score, _ = getattr(self, f"patchmatch_{li}")(
+                    ref_feature=ref_feature[f"stage_{li}"],
+                    src_features=src_features_l,
+                    ref_proj=ref_proj,
+                    src_projs=src_projs,
+                    depth_min=depth_min,
+                    depth_max=depth_max,
+                    depth=depth,
+                    img=getattr(self, f"imgs_{li}_ref"),
+                    view_weights=view_weights,
+                )
+
+            del src_features_l, ref_proj, src_projs, projs_l
+
+            depth_patchmatch[f"stage_{li}"] = depth
+
+            depth = depth[-1].detach()
+            if li > 1:
+                # upsampling the depth map and pixel-wise view weight for next stage
+                depth = F.interpolate(depth, scale_factor=2, mode="nearest")
+                view_weights = F.interpolate(view_weights, scale_factor=2, mode="nearest")
+
+        # step 3. Refinement
+        depth = self.upsample_net(self.imgs_0_ref, depth, depth_min, depth_max)
+        refined_depth["stage_0"] = depth
+
+        del depth, ref_feature, src_features
+
+        if self.training:
+            return {
+                "refined_depth": refined_depth,
+                "depth_patchmatch": depth_patchmatch,
+            }
+
+        else:
+            num_depth = self.patchmatch_num_sample[0]
+            score_sum4 = 4 * F.avg_pool3d(
+                F.pad(score.unsqueeze(1), pad=(0, 0, 0, 0, 1, 2)), (4, 1, 1), stride=1, padding=0
+            ).squeeze(1)
+            # [B, 1, H, W]
+            depth_index = depth_regression(
+                score, depth_values=torch.arange(num_depth, device=score.device, dtype=torch.float)
+            ).long()
+            depth_index = torch.clamp(depth_index, 0, num_depth - 1)
+            photometric_confidence = torch.gather(score_sum4, 1, depth_index)
+            photometric_confidence = F.interpolate(photometric_confidence, scale_factor=2, mode="nearest")
+            photometric_confidence = photometric_confidence.squeeze(1)
+
+            return {
+                "refined_depth": refined_depth,
+                "depth_patchmatch": depth_patchmatch,
+                "photometric_confidence": photometric_confidence,
+            }
+
+
+def patchmatchnet_loss(
+    depth_patchmatch: Dict[str, torch.Tensor],
+    refined_depth: Dict[str, torch.Tensor],
+    depth_gt: Dict[str, torch.Tensor],
+    mask: Dict[str, torch.Tensor],
+) -> torch.Tensor:
+    """Patchmatch Net loss function
+
+    Args:
+        depth_patchmatch: depth map predicted by patchmatch net
+        refined_depth: refined depth map predicted by patchmatch net
+        depth_gt: ground truth depth map
+        mask: mask for filter valid points
+
+    Returns:
+        loss: result loss value
+    """
+    stage = 4
+
+    loss = torch.Tensor([0])
+    for li in range(1, stage):

```

Comment:
I would keep the stage index as `l`, not as `li`, since we use the variable `mask_l` and `depth_gt_l` below